src/main/java/blusunrize/immersiveengineering/common/crafting/fluidaware/AbstractFluidAwareRecipe.java

```diff
@@ -0,0 +1,129 @@
+ * Details can be found in the license file in the root folder of this project
+ *
+ */
+
+package blusunrize.immersiveengineering.common.crafting.fluidaware;
+
+import blusunrize.immersiveengineering.common.crafting.IngredientFluidStack;
+import blusunrize.immersiveengineering.common.crafting.fluidaware.AbstractFluidAwareRecipe.IMatchLocation;
+import com.google.common.base.Preconditions;
+import net.minecraft.inventory.CraftingInventory;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.ICraftingRecipe;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.util.NonNullList;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.world.World;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public abstract class AbstractFluidAwareRecipe<MatchLocation extends IMatchLocation> implements ICraftingRecipe
+{
+	protected final static boolean[] BOOLEANS = {true, false};
+
+	private final NonNullList<Ingredient> recipeItems;
+	private final ItemStack recipeOutput;
+	private final ResourceLocation id;
+	private final String group;
+
+	public AbstractFluidAwareRecipe(
+			ResourceLocation idIn, String groupIn, NonNullList<Ingredient> recipeItemsIn, ItemStack recipeOutputIn
+	)
+	{
+		this.id = idIn;
+		this.group = groupIn;
+		this.recipeItems = recipeItemsIn;
+		this.recipeOutput = recipeOutputIn;
+	}
+
+	@Nonnull
+	@Override
+	public ResourceLocation getId()
+	{
+		return this.id;
+	}
+
+	@Nonnull
+	@Override
+	public String getGroup()
+	{
+		return this.group;
+	}
+
+	@Nonnull
+	@Override
+	public ItemStack getRecipeOutput()
+	{
+		return this.recipeOutput;
+	}
+
+	@Nonnull
+	@Override
+	public NonNullList<Ingredient> getIngredients()
+	{
+		return this.recipeItems;
+	}
+
+	@Override
+	public boolean matches(@Nonnull CraftingInventory inv, @Nonnull World worldIn)
+	{
+		return findMatch(inv)!=null;
+	}
+
+	@Nullable
+	protected abstract MatchLocation findMatch(CraftingInventory inv);
+
+	@Nonnull
+	@Override
+	public ItemStack getCraftingResult(@Nonnull CraftingInventory inv)
+	{
+		return this.getRecipeOutput().copy();
+	}
+
+	@Nonnull
+	@Override
+	public NonNullList<ItemStack> getRemainingItems(@Nonnull CraftingInventory inv)
+	{
+		NonNullList<ItemStack> remaining = NonNullList.withSize(inv.getSizeInventory(), ItemStack.EMPTY);
+		final MatchLocation offset = findMatch(inv);
+		Preconditions.checkNotNull(offset);
+
+		for(int x = 0; x < inv.getWidth(); ++x)
+			for(int y = 0; y < inv.getHeight(); ++y)
+			{
+				final int invIndex = getInventoryIndex(inv, x, y);
+				final int ingrIndex = offset.getListIndex(x, y);
+				if(ingrIndex >= 0)
+				{
+					Ingredient ingr = getIngredients().get(ingrIndex);
+					final ItemStack item = inv.getStackInSlot(invIndex);
+					ItemStack result = ItemStack.EMPTY;
+					if(ingr instanceof IngredientFluidStack)
+						result = ((IngredientFluidStack)ingr).getExtractedStack(item).copy();
```

Comment:
I am not positive but I believe it would make sense to copy the item here (or in getExtractedStack before extracting), just so that if a crafting implementer queries what remaining items are for a recipe there are no side effects of the container being drained just because of the getter being called to say see if an automated crafter has room to also store the remaining stacks.